Fix the support for WinRT/UWP

vnen opened the pull request:
~~**Note:** much of the commit history can be discarded, so this PR should likely be squashed on merge. I used commits to save work and I'm not the most linear person in the world, so work is scattered and quite hard to group. I'll leave it like this for now though, since it might be easier to pinpoint specific changes if need be.~~ Nevermind, I rebased the history already,
## Building
- I'm using Visual Studio Community 2015 Update 3. May or may not work with earlier versions. Please report if you find that it works or not with other versions.
- C++ tools for VS are needed of course, and also the Windows 10 SDK (all of which can be selected in the VS installer).
- Clone the default `ms-master` branch from the [Microsoft ANGLE fork](https://github.com/Microsoft/angle) somewhere.
- Set the `ANGLE_SRC_PATH` environment variable to the ANGLE clone full path.
- Same instructions as [Windows](http://docs.godotengine.org/en/latest/reference/compiling_for_windows.html): open the correct prompt (can also be the ARM prompt to build for devices). Run `scons p=winrt target=release_debug`.
  - This will also build ANGLE for the correct architecture, so no need to pre-build it.
### Templates

There's the `tools/dist/uwp_template` folder with the main template. Add the built `.exe` file here, rename it to `godot.winrt.exe` and Zip to a file called `winrt_%arch_%target`, where `%arch` is the architecture (`arm`, `x86` or `x64`) and `target` is either `release` or `debug`.

The template also need the built ANGLE DLLs. Copy `libEGL.dll` and `libGLESv2.dll` from `%ANGLE_SRC_PATH%\winrt\10\src\Release_%arch` (where `%arch` is `Win32`, `x64`, or `ARM`) to the root of the zip template.

As the other platforms, you need _all_ templates installed to be able to export. Those don't need to be correct, just to exist in the templates folder.

Note that the emulator only runs the `x86` architecture.
## Exporting

The exporter generate an **unsigned** `.appx` package. One important thing to change is the `Publisher` option in the export settings to match the certificate you'll use. Packages cannot be installed without being signed by a trusted certificate.
### Signing

There's code for generating the signature but the digests don't match. It was ported from [fb-util-for-appx](https://github.com/facebook/fb-util-for-appx) and I failed to see what is the difference. I left the code because it's almost working and I'm willing to revisit it in the future.

Hence, signing can only be done in Windows for now (note you don't need to sign beforehand when submitting to the store). Instructions for [making a certificate](https://msdn.microsoft.com/en-us/library/windows/desktop/jj835832%28v=vs.85%29.aspx) and [signing a package](https://msdn.microsoft.com/en-us/library/windows/desktop/jj835835%28v=vs.85%29.aspx) are provided by Microsoft.
### Bundles

It's theoretically possible to make "App Bundles". Those are similar to "fat" binaries, i.e. they contain packages for multiple architectures. There's no code for that now, but it should be easy to do by refactoring the packager code.
## Third-party
### ANGLE

~~I added a module for [ANGLE](http://angleproject.org) since it's needed for UWP. Following the pattern already established in Godot to add the code and statically link instead of providing DLLs. Also, ANGLE may be used for Windows Desktop (which may solve some issues with problematic graphic cards) and maybe to EGL contexts like the Raspberry Pi.~~

No ANGLE module anymore. Now there's a dependency on external ANGLE source, as described in the "Building" section.
### OpenSSL

Microsoft have their own [fork of OpenSSL](https://github.com/Microsoft/openssl) to make it work for WinRT. Luckily the changes I ported are quite minimal and to make it easier I provided a patch file to see what was changed and apply it to updates.

From the MS fork I also took a file with shim code for unsupported API under UWP. This is only an _addition_ so it shouldn't cause problems.
### Enet

I also needed to apply a patch to make enet build. For this I did not store the patch file, but the changes were the same as [enet-uwp](https://github.com/ahrnbom/enet-uwp/commit/c14968f8f58cc96c5054e0c1a396701586582dbe). Those might even be pushed upstream, but AIUI enet is not really being maintained and not accepting PRs.
## Missing features

I don't see any important thing missing but I might have overlooked something. Nevertheless, here it's a list of things that wasn't ported because it's either too difficult or even impossible:
- Window mode (resizable/borderless).
- Maximizing/minimizing the window.
- Monitor info (amount/size). This _might_ be possible with DirectX though.
- <kbd>Alt</kbd> key and any combination is not reported. The <kbd>AltGr</kbd> is reported as <kbd>Ctrl</kbd>+<kbd>Alt</kbd>.
- <kbd>F10</kbd> key is also not reported. Along with <kbd>Alt</kbd>, it's a System Key, which does not send events for the underlying window.
- Deploying and launching the app with one-click. 
  - The [Deployment namespace](https://msdn.microsoft.com/en-us/library/windows/apps/windows.management.deployment.aspx) has functionality to deployment, but I couldn't include the required header without getting a lot of redefinition errors.
  - Maybe it's possible to interface with prompt/PowerShell and use the commands to deploy and launch apps.
  - It's also pretty much impossible to detect the Windows 10 version without a manifest. Since the last trial for that failed for some people (#5364), it might be hard to make the feature version-wise (since you can't install UWP apps in earlier Windows versions).
## Extra tasks
- ANGLE can be integrated with Windows Desktop and maybe for other platforms too.
- XAudio2 audio driver might also be used for Windows Desktop (done in #6855).

akien-mga commented:
Awesome work! I'll allow myself a small `needs rebase`, as I would prefer not to merge 69 commits, but I don't to squash them all together as the diff would be *huge* and unreadable.
Especially things like the addition of ANGLE should stay in an own commit. I'll see during review if I see a pattern for grouping commits together without too much rebase pain :)

punto- commented:
This is awesome, thanks! One thing I was thinking about is, do we put the
template on tools, or on the platform? I think I moved the ios template to
platform/iphone at some point, just to have everything that belongs to a
platform inside its platform/ directory. This one is not a big deal tho
since it's an open platform (just like we have drivers/windows and not
something inside platform/windows)

I'm so happy it's a lot easier to build an launch than my old port, that
was a pain in the ass :p

On 1 September 2016 at 18:39, Rémi Verschelde <notifications@github.com>
wrote:

> Awesome work! I'll allow myself a small needs rebase, as I would prefer
> not to merge 69 commits, but I don't to squash them all together as the
> diff would be *huge* and unreadable.
> Especially things like the addition of ANGLE should stay in an own commit.
> I'll see during review if I see a pattern for grouping commits together
> without too much rebase pain :)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6363#issuecomment-244154342>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPV79NJxzdhs2AggFbZ-stgi12usaks5qlw3CgaJpZM4Jy6_G>
> .
>

akien-mga commented:
About the template folder, I'm actually about to make a PR to cleanup `tools/` a bit. We can discuss there if we want to put them someplace else than what I'm proposing (in my PR `tools/dist`).

reduz commented:
I'm against bundling Angle inside Godot, it's a huge library, I think we
should have it as a third party dependency for the platform

On Thu, Sep 1, 2016 at 5:19 PM, Rémi Verschelde <notifications@github.com>
wrote:

> About the template folder, I'm actually about to make a PR to cleanup
> tools/ a bit. We can discuss there if we want to put them someplace else
> than what I'm proposing (in my PR tools/dist).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6363#issuecomment-244199480>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2xqATRFUxC73OCXTRLVaCaaRGwkUks5qlzNegaJpZM4Jy6_G>
> .
>

vnen commented:
> I'm against bundling Angle inside Godot, it's a huge library, I think we
should have it as a third party dependency for the platform.

Well, it's a trade-off. ANGLE still must be built. So Godot can either integrate the sources and just build everything together or provide the 6 DLLs (2 for each architecture). Also, building the templates will require a build of ANGLE and a configuration for its path to get the includes and libraries, which will certainly make it harder to automate.

Another option is to keep the module but remove the third-party sources. When building the template then it's just a matter of dropping the source there before compiling.

reduz commented:
How does Angle provide their headers/DLLs?

On Thu, Sep 1, 2016 at 5:54 PM, George Marques <notifications@github.com>
wrote:

> I'm against bundling Angle inside Godot, it's a huge library, I think we
> should have it as a third party dependency for the platform.
>
> Well, it's a trade-off. ANGLE still must be built. So Godot can either
> integrate the sources and just build everything together or provide the 6
> DLLs (2 for each architecture). Also, building the templates will require a
> build of ANGLE and a configuration for its path to get the includes and
> libraries, which will certainly make it harder to automate.
>
> Another option is to keep the module but remove the third-party sources.
> When building the template then it's just a matter of dropping the source
> there before compiling.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6363#issuecomment-244209158>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2wl1kdpILArpstgrUfvYtU9Qr0Leks5qlzuLgaJpZM4Jy6_G>
> .
>

vnen commented:
@reduz:
> How does Angle provide their headers/DLLs?

They only provide the source code. Not even the Microsoft fork has distribution of binaries. The headers they have are the standard GLES/EGL provided by the Khronos Group.

reduz commented:
is it too difficult to get headers/dlls? because we could use an
environment var:

ANGLE_SDK_ROOT

pointing to a dir with headers/bin where you compiled it.

if not detected, angle will not build. Most platforms expect these kind of
things in Godot


On Thu, Sep 1, 2016 at 6:50 PM, George Marques <notifications@github.com>
wrote:

> @reduz <https://github.com/reduz>:
>
> How does Angle provide their headers/DLLs?
>
> They only provide the source code. Not even the Microsoft fork has
> distribution of binaries. The headers they have are the standard GLES/EGL
> provided by the Khronos Group.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6363#issuecomment-244223739>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z25FeZ4KReRD0bHD78vWKlbu03rhJks5ql0ikgaJpZM4Jy6_G>
> .
>

vnen commented:
I guess we can use the [Microsoft fork](https://github.com/Microsoft/angle) which already has the solutions ready to build. Maybe it's possible to use SCons to call a MSBuild command-line to compile ANGLE?

Since they have a standard location, it's possible to use a environment variable (which I was using before integrating the source).

Still, we'd have to provide export templates with DLLs. I'm not sure how to automate the build, but I guess it can clone the Git repo and build with MSBuild.

vnen commented:
I removed ANGLE sources and updated the build instructions.

vnen commented on platform/winrt/os_winrt.cpp:
```diff
@@ -482,28 +662,19 @@ OS::Time OSWinrt::get_time(bool utc) const {
-
-	ret.bias = info.Bias;
+	ret.bias = -3;
+	ret.name = "BRT";
```
Heh, did this just to make it build and them completely forgot to actually implement it...

vnen committed:
Implement missing WinRT functions

- Fix buildsystem for WinRT/UWP platform.
- Add audio driver and joystick mapping for WinRT.
- Enable thread class for WinRT.
- Refactor MSVC compiler architecture detection to methods.py, so it can
  be used by Windows and WinRT.

vnen committed:
Remove embedded ANGLE headers

There's now a dependency on external ANGLE code, those are not needed
anymore.

vnen committed:
Fix root path resolution on PCK file

vnen committed:
Fix drivers coding for WinRT

- Add a proper function to retrieve IP addresses.
- Solve issues with Windows FileAccess and DirAccess to use the same code
  for WinRT.
- Add patches to the GLES2 rasterizer to workaround ANGLE issues.

vnen committed:
Patch thirdy-party libraries to build for WinRT

- Patch enet code.
- Patch OpenSSL code and add shims for unavailable API.
- Add extra definition header for Freetype.

vnen committed:
Add Windows Universal export to editor

- Use OPENSSL_ENABLED definition to the whole source to detect it
  anywhere.
- Add WinRT/UWP template files with manifest and default images.

vnen commented:
Rebased it to 6 commits, @akien-mga

akien-mga commented:
@vnen It seems to fail building with gcc and clang (those are not false positive from travis):
```
platform/winrt/export/export.cpp:343:30: error: extra qualification 'AppxPackager::' on member 'buf_put_int16' [-fpermissive]
```

vnen commented:
Fixed the build issue.

vnen commented:
I don't know why it's failing now. It's something with the builtin OpenSSL but it works for the non-gcc builds. does anyone have a clue why there's an undefined reference to something that was defined?

vnen commented:
I just commented a snippet since it's about signing and not really in use right now. Will have to actually solve this weird linking issue in the future when working on signing.

vnen commented:
Moved the template to `tools/dist/uwp_template` to follow #6364.

Marqin commented:
@vnen , on Xbox One:

>Apps that exceed these memory requirements will encounter memory allocation failures. For more information about monitoring memory use, see the [MemoryManager class reference](https://msdn.microsoft.com/library/windows/apps/windows.system.memorymanager.aspx).

Does this PR support this properly?

vnen commented:
@Marqin there's no support for that. I can add, but I'm not sure what to do after hitting the memory limit. Also, I don't have a console to properly test it.

punto- commented:
I don't think the port specifically has to have support for that, other
than maybe some API in OS that tells you how much memory is left (which I
don't think we have at the moment). The memory allocator in the engine will
return ERR_OUT_OF_MEMORY when an allocation fails.. it's not a recoverable
state generally, you just have to avoid it.

On 13 September 2016 at 14:26, George Marques <notifications@github.com>
wrote:

> @Marqin <https://github.com/Marqin> there's no support for that. I can
> add, but I'm not sure what to do after hitting the memory limit. Also, I
> don't have a console to properly test it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6363#issuecomment-246757949>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPVz3j9JErFh97TlkoFqiHdODM22sks5qptzagaJpZM4Jy6_G>
> .
>
